File: java/Divide&Conqure/QuickSort.java
public class QuickSort {
    public static int partition(int arr[], int si, int ei) {
      int pivot = arr[ei];
      int i = si - 1; // to make place for element smaller than pivot

      for(int j=si; j<ei; j++){
        if(arr[j] <= pivot){
            i++;
            //swap
            int temp = arr[j];
            arr[j] = arr[i];
            arr[i] = temp;
         }
     }
         i++;
         //swap
         int temp = pivot;
         arr[ei] = arr[i];
         arr[i] = temp;

      return i;  // return pIdx
   }
   public static void quicksort(int arr[], int si, int ei) {
     // base case
     if(si >= ei){
        return;
     }
      // find pivot (last index)
      int pidx = partition(arr,si,ei);
      quicksort(arr, si, pidx-1);
      quicksort(arr, pidx+1, ei);
   }

   public static void printarr(int arr[]){
       for(int i = 0; i < arr.length; i++){
           System.out.print(arr[i] + " ");
       }
        System.out.println();
   }
   public static void main(String args[]){
       int arr[] = {6,3,9,5,2,8};
       quicksort(arr,0,arr.length-1);
       printarr(arr);
   }
}

File: c/sorting.c/QuickSort.c
#include<stdio.h>
    int partition(int arr[],int first,int last){
     int pivot=arr[last];
     int i=first-1;
     int j=first;
     int temp;
     for(j=first;j<last;j++){
      if(arr[j]<pivot){
        i++;
        // swap arr[i] and arr[j]
        temp=arr[i];
        arr[i]=arr[j];
        arr[j]=temp;
      }
     }  
         // swap arr[i+1] and arr[last]
        temp=arr[i+1];
        arr[i+1]=arr[last];
        arr[last]=temp;
        return i+1;
    }

  void quicksort(int arr[],int first,int last){
    if(first<last) {
     int pi=partition(arr,first,last);
     quicksort(arr,first,pi-1);
     quicksort(arr,pi+1,last); 
    }
  }
int main(){
    int arr[]={5,3,8,9,1,4};
    int n=sizeof(arr)/sizeof(int);
    int first=0;
    int last=n-1;
    quicksort(arr,0,n-1);
    for(int i=0;i<n;i++){
    printf("%5d",arr[i]);
    }
    return 0;
}
           
